added playground rules (#61)
dappforce/subsocial-docs: docs/develop/Playground.md

---
id: playground
title: Playground Rules
displayed_sidebar: developSidebar
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import CodeBlock from '@theme/CodeBlock';

> Subsocial Playground is an easy tool that allows you to run code snippets from our SDK without downloading anything.

Reviewer: ```suggestion
Subsocial's Playground is an easy tool that allows you to run code snippets from our SDK without downloading anything.
```

> The Playground provides a perfect environment to play, test and debug some of your code snippets before building a real project. 

Reviewer: ```suggestion
This provides a perfect environment to play, test and debug some of your code snippets before building a real project. 
```

:::info

> It's running on our Testnet called Soonsocial. You can get some test tokens from our [discord bot](https://discord.gg/JRTMTtPK).

Reviewer: ```suggestion
The Playground is running on our Testnet called Soonsocial. You can get some test tokens from our [Discord bot](https://discord.gg/JRTMTtPK).
```

:::

> ### Tryout the playground: [Subsocial Playground](https://docs.subsocial.network/playground).

Reviewer: ```suggestion
### Try out the [Subsocial Playground](https://docs.subsocial.network/playground).
```

> ## In-built APIs

Reviewer: ```suggestion
## Built-in APIs
```

It comes with a set of rules & in-built function to help you focus on library methods rather than writing everything from scratch. The following list represents these functions and their code: 

### Subsocial API

> The Subsocial Playground comes with Subsocial SDK's `api` instance. This is use to interact with most of the SDK methods, here's how it's being created underhood:

Reviewer: ```suggestion
The Subsocial Playground comes with the Subsocial SDK's `api` instance. This is used to interact with most of the SDK methods. Here is how it's being created under the hood:
```

<Tabs
  defaultValue="testnet"
  values={[
    {label: 'TestNet', value: 'testnet'},
    {label: 'MainNet', value: 'mainnet'},
    {label: 'LocalNet', value: 'localnet'},
  ]}>
  <TabItem value="mainnet">

```ts
  const configDetails = {
    substrateNodeUrl: 'wss://para.f3joule.space',
    ipfsNodeUrl: 'https://app.subsocial.network/ipfs'
  }
```

  </TabItem>
  <TabItem value="testnet">

```ts
  const configDetails = {
    substrateNodeUrl: 'wss://rco-para.subsocial.network',
    ipfsNodeUrl: 'https://staging.subsocial.network/ipfs'
  }
```

  </TabItem>
  <TabItem value="localnet">

```ts
  const configDetails = {
    substrateNodeUrl: 'http://127.0.0.1:9944',
    ipfsNodeUrl: 'http://127.0.0.1:8080'
  }
```
:::caution

> Make sure to run local Subsocial & IPFS node before using these configs in your project.

Reviewer: ```suggestion
Make sure to run local Subsocial & IPFS nodes before using these configs in your project.
```

:::
  </TabItem>
</Tabs>

```typescript
  const api = api = await SubsocialApi.create({
    ...configDetails,
    useServer: {
      httpRequestMethod: 'get'
    }
  })
```

### IPFS API

> The Subsocial Playground comes with CRUST IPFS API used as `ipfs` instance. This is use to store data related spaces, posts and comments on the network. Let's see how the configuration is set:

Reviewer: ```suggestion
The Subsocial Playground comes with the CRUST IPFS API's `ipfs` instance. This is use to store data related spaces, posts and comments on the network. Let's see how the configuration is set:
```

```typescript
  import { generateCrustAuthToken } from '@subsocial/api/utils/ipfs'

  const authHeader = generateCrustAuthToken('bottom drive obey lake curtain smoke basket hold race lonely fit walk//Alice')

  const ipfs = new SubsocialIpfsApi({
    ipfsNodeUrl: 'https://crustwebsites.net'
  })

  ipfs.setWriteHeaders({
    authorization: 'Basic ' + authHeader
  })
```

### Keyring

> The Subsocial Playground comes with a basic `keyring` instance to help you sign and send transactions. By default, we use the Alice (Test) account for our Testnet transactions but you can change it in our Playground. 

Reviewer: ```suggestion
The Subsocial Playground comes with a basic `keyring` instance to help you sign and send transactions. By default, we use the Alice (Test) account for our Testnet transactions, but you can change it if you want. 
```

```typescript
  import { Keyring } from "@polkadot/api"
  import { waitReady } from '@polkadot/wasm-crypto';

  await waitReady()
  const keyring = new Keyring({ type: 'sr25519' })
```

### Sending Transaction

> The Subsocial Playground allows you to sign and send the transaction through Mnemonic using Keyring and Polkadot.js extension both. 

Reviewer: ```suggestion
The Subsocial Playground allows you to sign and send transactions with Mnemonic phrases using both Keyring and the Polkadot.js extension. 
```

#### Using Keyring

> To sign and send transactions using Keyring through Mnemonic, use:

Reviewer: ```suggestion
To sign and send transactions with Mnemonic phrases through Keyring, use:
```

```typescript
  transaction.signAndSend(pair, logger) 
```

The logger is a method that tracks the transaction and show toasts in UI. Here's how it works:

```typescript
const logger = (result: any) => {
    const { status } = result

    if (!result || !status) {
      return;
    }
    if (status.isFinalized) {
      const blockHash = status.isFinalized
        ? status.asFinalized
        : status.asInBlock;
      console.log('✅ Tx finalized. Block hash', blockHash.toString());
      showToast(`✅ Transaction: ${status.isFinalized ? 'Finalised' : 'Added in Block'}`);
      const newIds = getNewIdsFromEvent(result); // get first argument from array.
      if (newIds.length > 0) {
        showToast(`⚡️ New Item Id: ${newIds[0]}`)
      }
      return;
    } else if (result.isError) {
      console.log(JSON.stringify(result));
      showToast(JSON.stringify(result));
    } else {
      console.log('⏱ Current tx status:', status.type);
      showToast(`⏱ Current tx status: ${status.type}`);
    }
  }
```

#### Using Polkadot.js Extension

> Signing using the polkadot.js extension: 

Reviewer: ```suggestion
Signing using the Polkadot.js extension: 
```
